### ISSUE TITLE
Format after RustExpandMacro

### ISSUE THREAD

**Issue by sharksforarms** [simrat39/rust-tools.nvim#2]
It would be nice to be able to format the code in the expanded, I've tried running `lua vim.lsp.buf.formatting_sync()` but didn't affect the code

**Comment by simrat39** [comment]
I'm sorry but this isn't really possible due to a couple of reason:
1. The LSP doesn't attach itself to the scratch (throwaway) buffer created while showing the expansion
2. The macro expansion sent by rust analyzer isn't even valid rust syntax I'm pretty sure (has a $ sign and brackets and what not) so rust-analyzer can't format it.
3. Need to make edits to the expansion to make the LSP even consider it for formatting

**Comment by simrat39** [comment]
Actually, this might be possible if we format it in the code (not relying on the lsp)
![image](https://user-images.githubusercontent.com/37233846/113799227-8b26ae80-9709-11eb-8c09-9631847455a2.png)
heres something i was able to get with just replacing commas with commas + new lines + spaces

**Comment by simrat39** [comment]
This doesn't look to bad :thinking: :thinking: , what do you thing @sharksforarms

**Comment by sharksforarms** [comment]
Hm not sure. I think it's fair enough to display exactly what rust-analyzer is giving and if possible, use the LSP to format it (as an option)

As you can see below, everything is still readable... you're right though, some things get messed up. For example the lifetime `'a` get's squashed  with `BitSlice` as `'aBitSlice` which makes this no be able to format properly. If I fix those, I can run rustfmt on the output

Default:
![image](https://user-images.githubusercontent.com/10912917/113800692-64767100-9725-11eb-89b0-83c659ec2d60.png)

Formatted (after fixing lifetime):
![image](https://user-images.githubusercontent.com/10912917/113801019-01d1a500-9726-11eb-8e5a-637ef77b9075.png)

**Comment by sharksforarms** [comment]
Related: https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/expand_macro.rs#L36-L38

**Comment by simrat39** [comment]
> Related:
> https://github.com/rust-analyzer/rust-analyzer/blob/cf41e1410a0cf87a05b2dd632e3b5d67f8d55327/crates/ide/src/expand_macro.rs#L36-L38

Fair, lets wait for rust-analyzer to get it sorted, till then lets keep this issue open

**Comment by simrat39** [comment]
@sharksforarms is it fine now? I'll test it out tomorrow, need to start using ra nightly, anyways good work u fixed it so fast :+1:

**Comment by sharksforarms** [comment]
> @sharksforarms is it fine now? I'll test it out tomorrow, need to start using ra nightly, anyways good work u fixed it so fast +1

So the PR I made was just to fix the lifetime space issue as described above. I did reach out to RA in their [zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Format.20after.20Expand.20Macro) and they did mention an internal formatting library was in the making. Probably better to wait :+1: Alternatively, rustfmt could be ran over the output or if there's a way to attach the LSP we could try that? This is where my knowledge of neovim get's fuzzy
